container_test.go

```diff
@@ -324,6 +325,113 @@ func Test_GetLogsFromFailedContainer(t *testing.T) {
 	}
 }
 
+type dockerImageSubstitutor struct{}
+
+func (s dockerImageSubstitutor) Description() string {
+	return "DockerImageSubstitutor (prepends docker.io)"
+}
+
+func (s dockerImageSubstitutor) Substitute(image string) (string, error) {
+	return "docker.io/" + image, nil
+}
+
+// noopImageSubstitutor {
+type NoopImageSubstitutor struct{}
+
+// Description returns a description of what is expected from this Substitutor,
+// which is used in logs.
+func (s NoopImageSubstitutor) Description() string {
+	return "NoopImageSubstitutor (noop)"
+}
+
+// Substitute returns the original image, without any change
+func (s NoopImageSubstitutor) Substitute(image string) (string, error) {
+	return image, nil
+}
+
+// }
+
+type errorSubstitutor struct{}
+
+var errSubstitution = errors.New("substitution error")
+
+// Description returns a description of what is expected from this Substitutor,
+// which is used in logs.
+func (s errorSubstitutor) Description() string {
+	return "errorSubstitutor"
+}
+
+// Substitute returns the original image, but returns an error
+func (s errorSubstitutor) Substitute(image string) (string, error) {
+	return image, errSubstitution
+}
+
+func TestImageSubstitutors(t *testing.T) {
+	tests := []struct {
+		name          string
+		image         string // must be a valid image, as the test will try to create a container from it
+		substitutors  []ImageSubstitutor
+		expectedImage string
+		expectedError error
+	}{
+		{
+			name:          "No substitutors",
+			image:         "alpine",
+			expectedImage: "alpine",
+		},
+		{
+			name:          "Noop substitutor",
+			image:         "alpine",
+			substitutors:  []ImageSubstitutor{NoopImageSubstitutor{}},
+			expectedImage: "alpine",
+		},
+		{
+			name:          "Prepend namespace",
+			image:         "alpine",
+			substitutors:  []ImageSubstitutor{dockerImageSubstitutor{}},
+			expectedImage: "docker.io/alpine",
+		},
+		{
+			name:          "Substitution with error",
+			image:         "alpine",
+			substitutors:  []ImageSubstitutor{errorSubstitutor{}},
+			expectedImage: "alpine",
+			expectedError: errSubstitution,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			ctx := context.Background()
+			req := ContainerRequest{
+				Image:             test.image,
+				ImageSubstitutors: test.substitutors,
```

Comment:
I wonder if `ImageSubstitutors` should apply to `Image` instead of the `ContainerRequest`.

Reply:
Image is a string. Is that what you mean?

There is a `WithImageSubstitutors` functional option that will be used to build the request. For modules, the API is more "elegant" nowadays, for the genetic container, before V1, it's part of the container request.